### ISSUE TITLE
okhttp-sse RealCall CastException

### ISSUE THREAD

**Issue by ywang-nr** [square/okhttp#7724]
**Summary of the problem**

1. New Relic is trying to add support for okhttp-sse and found out one part of the code for RealCall causes cast exception.
> The New Relic Android agent provides performance monitoring instrumentation, we use ASM to perform byte-code injection to collect data from performance issues to tiny errors within the code.

> When integrate both New Relic android agent and okhttp-sse library, we found a crash in `RealEventSource` class file at this line--> `call = client.newCall(request) as RealCall`

**Why are we getting the crash and potential solutions:**
1. RealCall cannot be implemented as it's a final class - Solution 1: Can we adjust RealCall class to be public?
2. Casting newCall as RealCall causes the crash `call = client.newCall(request) as RealCall` - Solution 2: Can we adjust this part of the casting to use a public class?

Above are what I have found out so far, any insights are appreciated. Thank you.

**Comment by yschimke** [comment]
Is there a New Relic bug, or PR you can point us at? Or a stack trace for the class cast exception?

Is it this? https://github.com/newrelic/newrelic-java-agent/blob/3e19dce8fc731279583e8507e5ff47dd9bea0ab6/instrumentation/okhttp-4.4.0/src/main/java/com/nr/agent/instrumentation/okhttp44/internal/RealCall_Instrumentation.java

On 2), I'm a bit sympathetic to OkHttp itself not assuming newCall returns RealCall https://github.com/square/okhttp/blob/okhttp_4x/okhttp-sse/src/main/kotlin/okhttp3/internal/sse/RealEventSource.kt#L41 but I'm not sure what the right fix would be.

I think having a Call.Factory that returns other call implementations is a supported part of the API.  In this case we use OkHttp directly, so casting was considered safe.  But I think we should fix this.

But it would be good to work through this, we aren't against monitoring tools. One way might be to have a public API that let's us unwrap types like this. Say DelegatedCall, with an unwrap method?

In terms of making RealCall public, this is precisely why the API is controlled, we are strict about not breaking our public API, and tools that use internal APIs will break frequently, because we do change those internals. It's been vital in letting us evolve the project.  So it won't be made public API or non final. 

cc @swankjesse @JakeWharton thoughts?

**Comment by ywang-nr** [comment]
Thank you for taking the time to review this. 

**To answer your questions first:**
1. There isn't a bug or PR I can point to you at the moment, but I can provide a sample app to reproduce the issue. Sample app and logs are attached. 
2. The linked [PR](https://github.com/newrelic/newrelic-java-agent/blob/3e19dce8fc731279583e8507e5ff47dd9bea0ab6/instrumentation/okhttp-4.4.0/src/main/java/com/nr/agent/instrumentation/okhttp44/internal/RealCall_Instrumentation.java) you mentioned is a great find, that is our Java agent, it's similar and I've already looked into it before reaching out, that approach won't be able to apply to our android mobile agent.
3. The approach you're proposing is almost exactly what we're thinking, it could work if we can unwrap it to be a general public "Call" API

**Steps to reproduce:**
1. The exception will happen at line 86 in `MainActivity.kt` when it's trying to connect() in [RealEventSource](https://github.com/square/okhttp/blob/okhttp_4x/okhttp-sse/src/main/kotlin/okhttp3/internal/sse/RealEventSource.kt#L41)
> ```
EventSources.createFactory(client)
            .newEventSource(request = request, listener = eventSourceListener)
```
2. With New Relic's android library, we are transforming original `newCall()` to `com.newrelic.agent.android.instrumentation.okhttp3.CallExtension` which implements `Call` during build time, but our library never transform or be able to transform `RealCall` as it's a final class.

Hope this makes more sense and I am happy to schedule a meeting to demo and discuss how we both can do to approach this. 


[SSEDemoApp.zip](https://github.com/square/okhttp/files/10855293/SSEDemoApp.zip)

**Comment by yschimke** [comment]
Yep, I think a general rule would be any public decorator APIs should be safely unwrappable. OkHttp should support you applying these decorators and using as normal.

I'll put up a PR for discussion against 5.x and see where it goes. If we can agree then backport to a 4.x

**Comment by yschimke** [comment]
The only unfortunate bit is that we can't easily change the SSE API to Call.Factory for normal users of this API. So it only works in the world where OkHttpClient.newCall is weaved by your code. 

https://github.com/square/okhttp/blob/okhttp_4x/okhttp-sse/src/main/kotlin/okhttp3/internal/sse/RealEventSource.kt#L41

```
  fun connect(client: OkHttpClient) {
    val client = client.newBuilder()
        .eventListener(EventListener.NONE)
        .build()
    call = client.newCall(request) as RealCall
```

**Comment by yschimke** [comment]
experimenting on https://github.com/square/okhttp/pull/7728/files

**Comment by yschimke** [comment]
closing as no plan to fix
